ludwig/api.py

```diff
@@ -983,31 +994,90 @@ def generate(
         # warning, e.g.:
         # "A decoder-only architecture is being used, but right-padding was detected! For correct generation results, "
         # "please set `padding_side='left'` when initializing the tokenizer.
-        if not self.model.model.config.is_encoder_decoder:
-            padding_side = "left"
-        else:
-            padding_side = "right"
+        padding_side = "left" if not self.model.model.config.is_encoder_decoder else "right"
         tokenizer = HFTokenizer(self.config_obj.base_model, padding_side=padding_side)
 
         with self.model.use_generation_config(generation_config):
             start_time = time.time()
-            inputs = tokenizer.tokenizer(input_strings, return_tensors="pt", padding=True)
-            input_ids = inputs["input_ids"].to("cuda")
-            attention_mask = inputs["attention_mask"].to("cuda")
+            tokenized_inputs = tokenizer.tokenizer(input_strings, return_tensors="pt", padding=True)
+            input_ids = tokenized_inputs["input_ids"].to("cuda")
+            attention_mask = tokenized_inputs["attention_mask"].to("cuda")
+            streamer = create_text_streamer(tokenizer.tokenizer) if streaming else None
+
+            if streamer:
+                outputs = self._generate_streaming_outputs(input_strings, input_ids, attention_mask, streamer)
+            else:
+                outputs = self._generate_non_streaming_outputs(input_strings, input_ids, attention_mask, streamer)
+
+            decoded_outputs = tokenizer.tokenizer.batch_decode(outputs, skip_special_tokens=True)
+            logger.info(f"Finished generating in: {(time.time() - start_time):.2f}s.")
+
+            return decoded_outputs[0] if len(decoded_outputs) == 1 else decoded_outputs
+
+    def _generate_streaming_outputs(
+        self,
+        input_strings: Union[str, List[str]],
+        input_ids: torch.Tensor,
+        attention_mask: torch.Tensor,
+        streamer: Union[TextStreamer, None],
```

Comment:
Very minor -- if we are already calling this method, then `streamer` should not be `None`...  Is this right?